File: src/main/java/org/ideplugins/vale_cli_plugin/activity/ValeStartupActivity.java
package org.ideplugins.vale_cli_plugin.activity;

import com.intellij.ide.BrowserUtil;
import com.intellij.ide.plugins.IdeaPluginDescriptor;
import com.intellij.ide.plugins.PluginManagerCore;
import com.intellij.notification.*;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.editor.EditorFactory;
import com.intellij.openapi.extensions.PluginId;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.startup.StartupActivity;
import com.intellij.psi.search.FilenameIndex;
import com.intellij.psi.search.GlobalSearchScope;
import org.apache.commons.lang3.StringUtils;
import org.ideplugins.vale_cli_plugin.listener.FileSavedListener;
import org.ideplugins.vale_cli_plugin.service.ValeCliExecutor;
import org.ideplugins.vale_cli_plugin.settings.ValeCliPluginConfigurationState;
import org.ideplugins.vale_cli_plugin.settings.ValePluginSettingsState;
import org.ideplugins.settings.SettingsProvider;
import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.Optional;

import static org.ideplugins.vale_cli_plugin.Constants.*;


public class ValeStartupActivity implements StartupActivity {

    private static void getValeFilesCount(Project project) {
        if (project.isDisposed()) {
            return;
        }
        ValePluginSettingsState settingsState = ValePluginSettingsState.getInstance();

        if (StringUtils.isNotBlank(settingsState.extensions)) {
            ValeCliExecutor executor = project.getService(ValeCliExecutor.class);
            GlobalSearchScope scope = GlobalSearchScope.projectScope(project);
            ApplicationManager.getApplication().runReadAction(() -> {
                int numberOfFiles = Arrays.stream(settingsState.extensions.split(",")).map(extension ->
                        FilenameIndex.getAllFilesByExt(project, extension, scope).size()).reduce(Integer::sum).orElse(0);
                executor.setNumberOfFiles(numberOfFiles);
            });
        }
    }

    private static void showUpdateNotification(Project project, IdeaPluginDescriptor pluginDescriptor,
                                               ValeCliPluginConfigurationState pluginSettings) {
        ApplicationManager.getApplication().invokeLater(() -> Optional.ofNullable(NotificationGroupManager.getInstance()
                .getNotificationGroup(NOTIFICATION_GROUP)).ifPresent(group -> {
            NotificationAction action =
                    NotificationAction.createSimple(UPDATE_NOTIFICATION_BODY, () -> BrowserUtil.browse(JB_MARKETPLACE_URL));
            Notification notification = group.createNotification(
                    UPDATE_NOTIFICATION_TITLE,
                    "",
                    NotificationType.INFORMATION).addAction(action);
            Notifications.Bus.notify(notification, project);
            pluginSettings.setLastVersion(pluginDescriptor.getVersion());
        }));
    }

    @Override
    public void runActivity(@NotNull Project project) {
        ApplicationManager.getApplication().executeOnPooledThread(() -> getValeFilesCount(project));
        FileSavedListener listener = FileSavedListener.getInstance(project);
        EditorFactory.getInstance().getEventMulticaster().addDocumentListener(listener, listener);
        listener.activate();
        PluginId id = PluginId.getId(PLUGIN_ID);
        IdeaPluginDescriptor pluginDescriptor = PluginManagerCore.getPlugin(id);
        if (pluginDescriptor != null) {
            ValeCliPluginConfigurationState pluginSettings = ApplicationManager.getApplication().getService(ValeCliPluginConfigurationState.class);
            String lastKnownVersion = pluginSettings.getLastVersion();

            pluginSettings.setSentryDsn(SettingsProvider.getInstance().getSentryUrl(id.getIdString()));

            if (!lastKnownVersion.isEmpty() && !lastKnownVersion.equals(pluginDescriptor.getVersion())) {
                showUpdateNotification(project, pluginDescriptor, pluginSettings);
            }
        }
    }
}

File: src/main/kotlin/org/ideplugins/vale_cli_plugin/activity/ValeStartupActivity.kt
package org.ideplugins.vale_cli_plugin.activity

import com.intellij.execution.ExecutionException
import com.intellij.execution.process.CapturingProcessAdapter
import com.intellij.execution.process.ProcessEvent
import com.intellij.ide.plugins.PluginManagerCore.getPlugin
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.application.invokeLater
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.extensions.PluginId
import com.intellij.openapi.project.Project
import com.intellij.openapi.startup.ProjectActivity
import org.ideplugins.vale_cli_plugin.Constants
import org.ideplugins.vale_cli_plugin.service.ValeCliExecutor
import org.ideplugins.vale_cli_plugin.service.ValeLsConfigService
import org.ideplugins.vale_cli_plugin.settings.*
import org.ideplugins.vale_cli_plugin.utils.NotificationHelper

class ValeStartupActivity : ProjectActivity {

    private val logger = Logger.getInstance(ValeStartupActivity::class.java)

    override suspend fun execute(project: Project) {
        val notificationHelper = NotificationHelper(project)

        configureValePath()
        checkValeConfig(project)
        runSyncIfNeeded(project, notificationHelper)
        checkIfPluginWasUpdated(notificationHelper)
    }

    private fun checkValeConfig(project: Project) {
        val settings = ValePluginSettingsState.getInstance()
        if (settings.valePath.isNullOrEmpty() ) {
            return
        }
        val projectSettings = ValePluginProjectSettingsState.getInstance(project)
        if (!ApplicationManager.getApplication().isUnitTestMode) {
            if (projectSettings.valeSettingsPath.isEmpty() && projectSettings.rootIni.isEmpty()){
                logger.info("Running vale-ls config")
                val lsConfigService = ValeLsConfigService.getInstance(project)
                val configuration = lsConfigService.loadConfigurationPaths()
                projectSettings.rootIni = configuration.rootIni()
            } else {
                logger.info("Not running vale-ls config again")
            }
        } else {
            projectSettings.rootIni = ".vale.ini"
        }
    }

    private fun checkIfPluginWasUpdated(notificationHelper: NotificationHelper) {
        val id = PluginId.getId(Constants.PLUGIN_ID)
        val pluginDescriptor = getPlugin(id) ?: return

        val settings = ApplicationManager.getApplication()
            .getService(ValeCliPluginConfigurationState::class.java)

        if (
            settings.lastVersion.isNotEmpty() &&
            settings.lastVersion != pluginDescriptor.version
        ) {
            invokeLater {
                notificationHelper.showPluginWasUpdatedNotification()
                settings.lastVersion = pluginDescriptor.version
            }
        }
    }

    private fun configureValePath() {
        val settings = ValePluginSettingsState.getInstance()
        if (settings.valePath.isBlank()) {
            logger.info("Finding vale path")
            settings.valePath = OSUtils.findValeBinaryPath()
        }
    }

    private fun runSyncIfNeeded(project: Project, notificationHelper: NotificationHelper) {
        val projectSettings = ValePluginProjectSettingsState.getInstance(project)
        val settings = ValePluginSettingsState.getInstance()

        if (settings.valePath.isNullOrEmpty() || !projectSettings.runSyncOnStartup || projectSettings.rootIni.isEmpty()) {
            return
        }

        val cliExecutor = ValeCliExecutor.getInstance(project)

        try {
            cliExecutor.runSyncCommand(object : CapturingProcessAdapter() {
                override fun processTerminated(event: ProcessEvent) {
                    if (event.exitCode == 0) {
                        notificationHelper.notifySyncSuccess()
                    } else {
                        notificationHelper.notifySyncFailure(
                            cliExecutor.parseErrorProcessOutput(output).text
                        )
                    }
                }
            })
        } catch (e: ExecutionException) {
            logger.debug(e)
            notificationHelper.notifySyncFailure(e.message)
        }
    }
}